### ISSUE TITLE
pselect6 with NULL {sigmask*,masksize} struct for 6th + 7th args does not work?

### ISSUE THREAD

**Issue by egrimley** [DynamoRIO/dynamorio#3911]
This program does not seem to run correctly under DynamioRIO:
```
#include <sys/select.h>
#include <sys/syscall.h>
#include <unistd.h>

int main()
{
  fd_set fds;
  struct timespec ts;

  FD_ZERO(&fds);
  ts.tv_sec = 2;
  ts.tv_nsec = 0;
  return syscall(SYS_pselect6, 0, 0, 0, &fds, &ts, 0);
}
```

I think it's because the pselect6 code in `pre_system_call` does not check for the sigmask parameter being NULL.

**Comment by AssadHashmi** [comment]
#3896

**Comment by derekbruening** [comment]
Xref original #2759 which @hgreving2304 implemented -- thus assigning (unless @egrimley or @AssadHashmi plan to fix?  it is not assigned though).

**Comment by hgreving2304** [comment]
Interesting. The test didn't catch because the glibc wrapper seems to always pass a struct to the system  call, even with pselect(...., NULL). Seems like the kernel accepts the system call with NULL pointer, but we don't.

**Comment by derekbruening** [comment]
@AssadHashmi -- you implied above that OpenMPI invokes the syscall with NULL?  Is it doing a raw syscall then, given @hgreving2304's comment?  (Certainly DR should handle *any* syscall arg -- just trying to understand the OpenMPI issue and whether it does relate to this.)

**Comment by derekbruening** [comment]
Note that this is not really the *sigmask* parameter being NULL, this is the synthetic struct holding the sigmask pointer and the sigmask size being NULL.  That's why it is a little strange: it seems like passing NULL for that struct is an error and is not the same as passing NULL for a real argument.

**Comment by AssadHashmi** [comment]
> @AssadHashmi -- you implied above that OpenMPI invokes the syscall with NULL? Is it doing a raw syscall then, given @hgreving2304's comment? (Certainly DR should handle _any_ syscall arg -- just trying to understand the OpenMPI issue and whether it does relate to this.)

It's not a raw syscall AFAICT:
```
        res = select(nfds, sop->event_readset_out,                               
            sop->event_writeset_out, NULL, tv);
```
strace for our simple OpenMPI app shows it appearing as:
```
pselect6(1, [0], NULL, NULL, {5, 0}, NULL) = 0 (Timeout)
```
Our OpenMPI issue has been fixed by #3911.

**Comment by derekbruening** [comment]
So it must be that your glibc passes NULL for the 2-arg struct.
